File: task310-rest/src/main/java/com/example/task310rest/service/MarkService.java
package com.example.task310rest.service;

import com.example.task310rest.dto.request.MarkRequestTo;
import com.example.task310rest.dto.response.MarkResponseTo;
import com.example.task310rest.entity.Mark;
import com.example.task310rest.exception.ResourceNotFoundException;
import com.example.task310rest.mapper.MarkMapper;
import com.example.task310rest.repository.MarkRepository;
import com.example.task310rest.repository.TweetMarkRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.stream.Collectors;

/**
 * Сервис для работы с Mark
 */
@Service
@RequiredArgsConstructor
public class MarkService {
    
    private final MarkRepository markRepository;
    private final TweetMarkRepository tweetMarkRepository;
    private final MarkMapper markMapper;
    
    /**
     * Создать новую метку
     */
    public MarkResponseTo create(MarkRequestTo requestTo) {
        Mark mark = markMapper.toEntity(requestTo);
        Mark savedMark = markRepository.save(mark);
        return markMapper.toResponseTo(savedMark);
    }
    
    /**
     * Получить метку по ID
     */
    public MarkResponseTo getById(Long id) {
        Mark mark = markRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Mark", id));
        return markMapper.toResponseTo(mark);
    }
    
    /**
     * Получить метку по имени
     */
    public MarkResponseTo getByName(String name) {
        Mark mark = markRepository.findByName(name)
                .orElseThrow(() -> new ResourceNotFoundException("Mark with name='" + name + "' not found"));
        return markMapper.toResponseTo(mark);
    }
    
    /**
     * Получить все метки
     */
    public List<MarkResponseTo> getAll() {
        return markRepository.findAll().stream()
                .map(markMapper::toResponseTo)
                .collect(Collectors.toList());
    }
    
    /**
     * Обновить метку
     */
    public MarkResponseTo update(Long id, MarkRequestTo requestTo) {
        Mark existingMark = markRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Mark", id));
        
        existingMark.setName(requestTo.getName());
        
        Mark updatedMark = markRepository.update(existingMark);
        return markMapper.toResponseTo(updatedMark);
    }
    
    /**
     * Частичное обновление метки (PATCH)
     */
    public MarkResponseTo partialUpdate(Long id, MarkRequestTo requestTo) {
        Mark existingMark = markRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Mark", id));
        
        markMapper.updateEntityFromRequestTo(requestTo, existingMark);
        Mark updatedMark = markRepository.update(existingMark);
        return markMapper.toResponseTo(updatedMark);
    }
    
    /**
     * Удалить метку
     */
    public void delete(Long id) {
        if (!markRepository.existsById(id)) {
            throw new ResourceNotFoundException("Mark", id);
        }
        
        // Удаляем связи с твитами
        tweetMarkRepository.removeAllByMarkId(id);
        
        markRepository.deleteById(id);
    }
}

File: 351002/Bubich/t310/app/services/mark_service.py
from typing import List
from app.models.mark import Mark
from app.dto.requests.mark_request import MarkRequestTo
from app.dto.responses.mark_response import MarkResponseTo
from app.repository.in_memory_repository import InMemoryRepository
from app.exceptions.custom_exceptions import NotFoundError, ValidationError


class MarkService:
    def __init__(self, repository: InMemoryRepository):
        self.repository = repository

    def create(self, request: MarkRequestTo) -> MarkResponseTo:
        self._validate_request(request)
        mark = Mark(name=request.name)
        saved_mark = self.repository.save(mark)
        return self._to_response(saved_mark)

    def get_by_id(self, id: int) -> MarkResponseTo:
        mark = self.repository.find_by_id(id)
        if not mark:
            raise NotFoundError(f"Mark with id {id} not found")
        return self._to_response(mark)

    def get_all(self) -> List[MarkResponseTo]:
        marks = self.repository.find_all()
        return [self._to_response(m) for m in marks]

    def update(self, id: int, request: MarkRequestTo) -> MarkResponseTo:
        self._validate_request(request)
        existing_mark = self.repository.find_by_id(id)
        if not existing_mark:
            raise NotFoundError(f"Mark with id {id} not found")

        existing_mark.name = request.name
        updated_mark = self.repository.update(existing_mark)
        return self._to_response(updated_mark)

    def delete(self, id: int) -> None:
        if not self.repository.delete_by_id(id):
            raise NotFoundError(f"Mark with id {id} not found")

    def get_by_story_id(self, story_id: int, story_mark_repo: InMemoryRepository) -> List[MarkResponseTo]:
        # Здесь нужна связующая таблица story_mark, но для простоты пропустим
        return self.get_all()

    def _validate_request(self, request: MarkRequestTo):
        if not request.name or len(request.name) < 2 or len(request.name) > 32:
            raise ValidationError("Name must be between 2 and 32 characters")

    def _to_response(self, mark: Mark) -> MarkResponseTo:
        return MarkResponseTo(
            id=mark.id,
            name=mark.name
        )